### ISSUE TITLE
SecurityContext.getAllDeclaredCallerRoles() returns empty set in case of EJB

### ISSUE THREAD

**Issue by sameer-pandit** [javaee-security-spec/soteria#89]
SecurityContext.getAllDeclaredCallerRoles() however does not belong to latest spec. but returns correct list of roles in case of a servlet. This does not happen in case of a EJB (see this [test](https://github.com/javaee/glassfish/blob/master/appserver/tests/appserv-tests/devtests/security/soteria/app-securitycontext-auth-ejb/src/main/java/org/glassfish/soteria/test/TestEJB.java)). It would be nice if we have consistent behaviour for Servlet and EJB. 

I guess support needs to be added for handling EJBRoleRefPermission in [JACC.filterRoles()](https://github.com/javaee-security-spec/soteria/blob/master/impl/src/main/java/org/glassfish/soteria/authorization/JACC.java) along with WebRoleRefPermission.

**Comment by arjantijms** [comment]
You're right. Since Will removed the method from the latest spec, we didn't continue with this.

I half would have expected  that if the EJB was called from a web context (vs when called via remote EJB invocation), the WebRoleRefPermission would still be there and thus the list would not be empty, but probably that assumption is wrong then.

I'll try your suggestion, thx!

**Comment by arjantijms** [comment]
@sameer-pandit The actual fix was a bit more work, but basically, like you said, included checking for EJBRoleRefPermission. What made it a bit more work is that there's no portable way to get the `EJB-name` (https://stackoverflow.com/a/19054544/472792) and there's no automatic mapping of roles to the "global" context like in the Servlet container.

Getting the `EJB-name` should be done with a SPI somewhat later I think, but for now I just used a simple class check and a `toString` which happens to contain the `EJB-name` in GlassFish and Payara.

I did a PR with this fix, and added tests for the EJB role checks. The `getAllDeclaredCallerRoles()` is not tested for as its RI specific.

Thanks again so much for bringing this to our attention! :)
